components/script/dom/notification.rs

```diff
@@ -0,0 +1,545 @@
+            title,
+            body,
+            data,
+            dir,
+            image,
+            icon,
+            badge,
+            lang,
+            silent,
+            origin,
+            vibration_pattern,
+            timestamp,
+            renotify,
+            tag,
+            require_interaction,
+            actions,
+        }
+    }
+}
+
+impl NotificationMethods<crate::DomTypeHolder> for Notification {
+    /// <https://notifications.spec.whatwg.org/#constructors>
+    fn Constructor(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+    ) -> Fallible<DomRoot<Notification>> {
+        // step 1: Check global is a ServiceWorkerGlobalScope
+        if global.downcast::<ServiceWorkerGlobalScope>().is_some() {
+            return Err(Error::Type(
+                "Notification constructor cannot be used in ServiceWorkerGlobalScope.".to_string(),
+            ));
+        }
+
+        // step 2: Check options.actions must be empty
+        if !options.actions.is_empty() {
+            return Err(Error::Type(
+                "Actions are only supported for persistent notifications.".to_string(),
+            ));
+        }
+
+        // step 3: Create a notification with a settings object
+        let notification =
+            create_notification_with_settings_object(global, title, options, proto, can_gc)?;
+
+        // FIXME: Run step 5.1, 5.2 in parallel
+        // step 5.1: If the result of getting the notifications permission state is not "granted",
+        //           then queue a task to fire an event named error on this, and abort these steps.
+        // TODO: `get_descriptor_permission_state` seems always assume
+        //       we are in non-secure environment and return "Prompt"
+        let permission_state = get_notifications_permission_state(global);
+        if permission_state != NotificationPermission::Granted {
+            notification
+                .upcast::<EventTarget>()
+                .fire_event(atom!("error"), CanGc::note());
+            // TODO: abort steps
+        }
+        // TODO: step 5.2: Run the notification show steps for notification
+        // https://notifications.spec.whatwg.org/#notification-show-steps
+
+        Ok(notification)
+    }
+
+    /// <https://notifications.spec.whatwg.org/#dom-notification-permission>
+    fn GetPermission(global: &GlobalScope) -> Fallible<NotificationPermission> {
+        Ok(get_notifications_permission_state(global))
+    }
+
+    /// <https://notifications.spec.whatwg.org/#dom-notification-requestpermission>
+    fn RequestPermission(
+        global: &GlobalScope,
+        permission_callback: Option<Rc<NotificationPermissionCallback>>,
+    ) -> Rc<Promise> {
+        // Step 2: Let promise be a new promise in this’s relevant Realm.
+        // FIXME: not sure what CanGC should provide here
```

Comment:
You can have a CanGc value passed as an argument by adding a Notification entry to the list of interfaces at https://github.com/servo/servo/blob/main/components/script_bindings/codegen/Bindings.conf#L43 and using the `canGc` property.